rich/markdown.py

```diff
@@ -449,83 +470,107 @@ def __rich_console__(
+        inline_style_tags = self.inlines
         new_line = False
-        for current, entering in nodes:
-            node_type = current.t
-            if node_type in ("html_inline", "html_block", "text"):
-                context.on_text(current.literal.replace("\n", " "), node_type)
-            elif node_type == "linebreak":
-                if entering:
-                    context.on_text("\n", node_type)
+        _new_line_segment = Segment.line()
+
+        for token in self._flatten_tokens(tokens):
+            node_type = token.type
+            tag = token.tag
+
+            entering = token.nesting == 1
+            exiting = token.nesting == -1
+            self_closing = token.nesting == 0
+
+            if node_type == "text":
+                context.on_text(token.content, node_type)
+            elif node_type == "hardbreak":
+                context.on_text("\n", node_type)
             elif node_type == "softbreak":
-                if entering:
-                    context.on_text(" ", node_type)
-            elif node_type == "link":
-                if entering:
-                    link_style = console.get_style("markdown.link", default="none")
-                    if self.hyperlinks:
-                        link_style += Style(link=current.destination)
+                context.on_text(" ", node_type)
+            elif node_type == "link_open":
+                href = str(token.attrs.get("href", ""))
+                link_style = console.get_style("markdown.link", default="none")
+                if self.hyperlinks:
+                    link_style += Style(link=href)
                     context.enter_style(link_style)
                 else:
+                    context.stack.push(Link.create(self, token))
+            elif node_type == "link_close":
+                link_style = console.get_style("markdown.link", default="none")
+                if self.hyperlinks:
                     context.leave_style()
-                    if not self.hyperlinks:
-                        context.on_text(" (", node_type)
-                        style = Style(underline=True) + console.get_style(
-                            "markdown.link_url", default="none"
-                        )
-                        context.enter_style(style)
-                        context.on_text(current.destination, node_type)
-                        context.leave_style()
-                        context.on_text(")", node_type)
-            elif node_type in inlines:
-                if current.is_container():
-                    if entering:
-                        context.enter_style(f"markdown.{node_type}")
-                    else:
-                        context.leave_style()
                 else:
-                    context.enter_style(f"markdown.{node_type}")
-                    if current.literal:
-                        context.on_text(current.literal, node_type)
+                    element: Any = context.stack.pop()
```

Comment:
Can we not type this as the base class?

Reply:
It doesn't work because access attributes below which don't exist on the base class, only on `Link` (a subclass of the base MarkdownElement). This is most of the reason why I'd like to encapsulate all the stuff below inside `Link.__rich_console__`, it's cleaner _and_ would resolve this typing issue as we wouldn't have those attribute accesses :)

Reply:
Good point, forgot about that - have added that in